From 7df501d0af9055a8904433f6a18015f3d9837102 Mon Sep 17 00:00:00 2001
From: Hunt Xu <mhuntxu@gmail.com>
Date: Wed, 3 May 2017 18:38:17 +0800
Subject: [PATCH] l3_db: update GatewayInUseByFloatingIp check

With EayunStack floatingip mechanism, floatingip no longer depend on
router gateway.

As with EayunStack floatingip mechanism, the floatingip port will be set
up on hosts and thus its status will be ACTIVE. So we do the filter
using these ports' statuses to identify whether a router gateway is
actually in use by any floatingip.

Fixes: redmine #9990

Signed-off-by: Hunt Xu <mhuntxu@gmail.com>
---
 neutron/db/l3_db.py | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/neutron/db/l3_db.py b/neutron/db/l3_db.py
index f535ac90e..d5ccc1d91 100644
--- a/neutron/db/l3_db.py
+++ b/neutron/db/l3_db.py
@@ -342,9 +342,14 @@ class L3_NAT_dbonly_mixin(l3.RouterPluginBase):
         if not port_requires_deletion:
             return
         admin_ctx = context.elevated()
-
-        if self.get_floatingips_count(
-            admin_ctx, {'router_id': [router_id]}):
+        fip_qry = context.session.query(FloatingIP)
+        fip_qry = fip_qry.join(
+            models_v2.Port,
+            FloatingIP.floating_port_id == models_v2.Port.id)
+        fip_qry = fip_qry.filter(
+            models_v2.Port.status == l3_constants.PORT_STATUS_DOWN,
+            FloatingIP.router_id == router_id)
+        if fip_qry.all():
             raise l3.RouterExternalGatewayInUseByFloatingIp(
                 router_id=router_id, net_id=router.gw_port['network_id'])
         with context.session.begin(subtransactions=True):
-- 
2.12.2

